flink-mesos/src/test/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManagerTest.java

```diff
@@ -829,4 +839,45 @@ public void testClearStateAfterRevokeLeadership() throws Exception {
 			verify(rmServices.schedulerDriver).stop(true);
 		}};
 	}
+
+	/**
+	 * Tests that RM and TM calculate same slot resource profile.
+	 */
+	@Test
+	public void testCreateSlotsPerWorker() throws Exception {
```

Comment:
Can this code be deduplicated? I guess because the relevant production code in yarn/mesos RM constructor is the same. Maybe, at least some separate test class called from both or some common component to share the same concern between yarn/mesos RMs.

Reply:
It also came to my mind that whether these two test cases could share some common component. My concern is where should add this common component to, given that these two cases actually belong to two different modules (flink-yarn and flink-mesos). I think it might be even worse to just have one not self-contained component in one module and using it from two other modules. What do you think?

Reply:
@xintongsong I think it is not maintenance friendly to have a verbatim copy of `YarnResourceManagerTest#testCreateSlotsPerWorker()` for Mesos:

1. These tests should be kept in sync but this is not obvious to the developer
1. It looks like that we have to copy these tests once again for the Kubernetes ResourceManager. The development efforts are already in progress:
    * https://issues.apache.org/jira/browse/FLINK-9953
    * https://issues.apache.org/jira/browse/FLINK-9495

I wonder if: 
1. it is necessary to duplicate the entire test to be confident that `updateTaskManagerConfigAndCreateWorkerSlotProfiles()` is called in the constructor of the ResourceManager?
1. duplicating the invocations of `updateTaskManagerConfigAndCreateWorkerSlotProfiles()` in the ResourceManager implementations, which consequently requires us to duplicate tests, is a design flaw?

Maybe someone who was in involved in the design of this feature can chime in (cc: @StephanEwen)

Reply:
By the common component I meant enabling code reuse as a benefit. Is it not possible to well define it and test once?

Reply:
Thank you for the suggestion, @GJL. I think this does solve the problem. In that case we should also move `slotsPerWorker` to the common base class ResourceManager, which may not be used by StandaloneResourceManager but still can be shared by Yarn/Mesos/Kubernetes ResourceManagers. I'll make an update.